node/builder.go

```diff
@@ -0,0 +1,364 @@
+package node
+
+// Review: is a builder pattern with dot-chaining idiomatic in go?
+
+import (
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-filecoin/chain"
+	"github.com/filecoin-project/go-filecoin/clock"
+	"github.com/filecoin-project/go-filecoin/consensus"
+	"github.com/filecoin-project/go-filecoin/core"
+	"github.com/filecoin-project/go-filecoin/net"
+	"github.com/filecoin-project/go-filecoin/net/pubsub"
+	"github.com/filecoin-project/go-filecoin/plumbing"
+	"github.com/filecoin-project/go-filecoin/plumbing/cfg"
+	"github.com/filecoin-project/go-filecoin/plumbing/cst"
+	"github.com/filecoin-project/go-filecoin/plumbing/dag"
+	"github.com/filecoin-project/go-filecoin/plumbing/msg"
+	"github.com/filecoin-project/go-filecoin/plumbing/strgdls"
+	"github.com/filecoin-project/go-filecoin/porcelain"
+	"github.com/filecoin-project/go-filecoin/proofs/verification"
+	"github.com/filecoin-project/go-filecoin/repo"
+	"github.com/filecoin-project/go-filecoin/state"
+	"github.com/filecoin-project/go-filecoin/wallet"
+	"github.com/ipfs/go-bitswap"
+	bsnet "github.com/ipfs/go-bitswap/network"
+	bserv "github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-graphsync"
+	"github.com/ipfs/go-graphsync/ipldbridge"
+	gsnet "github.com/ipfs/go-graphsync/network"
+	gsstoreutil "github.com/ipfs/go-graphsync/storeutil"
+	"github.com/ipfs/go-hamt-ipld"
+	bstore "github.com/ipfs/go-ipfs-blockstore"
+	offline "github.com/ipfs/go-ipfs-exchange-offline"
+	offroute "github.com/ipfs/go-ipfs-routing/offline"
+	"github.com/ipfs/go-merkledag"
+	libp2p "github.com/libp2p/go-libp2p"
+	autonatsvc "github.com/libp2p/go-libp2p-autonat-svc"
+	circuit "github.com/libp2p/go-libp2p-circuit"
+	"github.com/libp2p/go-libp2p-core/host"
+	p2pmetrics "github.com/libp2p/go-libp2p-core/metrics"
+	"github.com/libp2p/go-libp2p-core/routing"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	dhtopts "github.com/libp2p/go-libp2p-kad-dht/opts"
+	libp2pps "github.com/libp2p/go-libp2p-pubsub"
+	rhost "github.com/libp2p/go-libp2p/p2p/host/routed"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pkg/errors"
+)
+
+// TODO: Create a `Builder` struct to keep track of the configuration
+
+// Config is a helper to aid in the construction of a filecoin node.
+// TODO: rename to `BuilderConfig`
+type Config struct {
+	BlockTime   time.Duration
+	Libp2pOpts  []libp2p.Option
+	OfflineMode bool
+	Verifier    verification.Verifier
+	Rewarder    consensus.BlockRewarder
+	Repo        repo.Repo
+	IsRelay     bool
+}
+
+// ConfigOpt is a configuration option for a filecoin node.
+type ConfigOpt func(*Config) error
+
+// New creates a new node.
+// TODO: rename to `(*NodeBuilder) Build() (*Node, error)`
```

Comment:
This one is a free function, not a method, so it would be something like `node.Build(ctx, opts...)`.

Reply:
Yes, you are correct. It currently is a free function. The signature was in line with the Review comment asking if dot chaining was idiomatic, in that case the last step on the chain would be: `.Build()`.